src/telemetry/miner_telemetry.erl

```diff
@@ -0,0 +1,294 @@
+                        Line,
+                        fun(Cnt) -> Cnt + 1 end,
+                        1,
+                        V
+                    )
+                end,
+                #{Line => 1},
+                M
+            )
+        end,
+        #{Severity => #{Line => 1}},
+        C
+    ),
+    {noreply, State#state{counts = NewCounts, logs = NewLogs}};
+handle_cast(Msg, State) ->
+    lager:info("unhandled cast ~p", [Msg]),
+    {noreply, State}.
+
+terminate(_Reason, State) ->
+    [lager:stop_trace(Trace) || Trace <- State#state.traces],
+    ok.
+
+maybe_send_telemetry(Json, _InConsensus, Height,
+                     #state{ always_send = true, sigfun = SigFun } = State) ->
+    case application:get_env(miner, telemetry_url) of
+        {ok, URL} ->
+            {_Pid, Mref} = do_send(URL, Json, SigFun),
+            State#state{mref = Mref, logs = [], counts = #{},
+                        last_sent_height=Height};
+        _ ->
+            lager:warning("telemetry always_send is true, but no url is specified. not sending"),
+            State#state{logs = [], counts = #{}}
+    end;
+maybe_send_telemetry(Json, true, Height, #state{sigfun = SigFun} = State) ->
+    case application:get_env(miner, telemetry_url) of
+        {ok, URL} ->
+            {_Pid, Mref} = do_send(URL, Json, SigFun),
+            State#state{mref = Mref, logs = [],
+                        counts = #{}, last_sent_height = Height};
+        _ ->
+            State#state{logs = [], counts = #{}}
+    end;
+maybe_send_telemetry(Json, false, Height,
+                     #state{last_sent_height=Last, sigfun=SigFun} = State) ->
+    case application:get_env(miner, telemetry_url) of
+        {ok, URL} ->
+            case (Height - Last) >= ?TELEMETRY_SEND_INTERVAL andalso
+                 rand:uniform() >= 0.50 of
+                true ->
+                    {_Pid, Mref} = do_send(URL, Json, SigFun),
+                    State#state{mref = Mref, logs = [],
+                                counts = #{}, last_sent_height = Height};
+                false ->
+                    State
+            end;
+        _ -> State
+    end.
+
+do_send(URL, #{ height := Height, version := Version, id := Id } = Json, SigFun) ->
+    Compressed = zlib:gzip(jsx:encode(Json)),
+    Signature = SigFun(Compressed),
+    spawn_monitor(fun() -> send_telemetry(Compressed, Signature, Id, Height, Version, URL)
+                  end).
+
+get_metadata(Key, M) ->
+    get_metadata(Key, M, undefined).
+
+get_metadata(Key, M, Default) ->
+    case lists:keyfind(Key, 1, M) of
+        {Key, Value} -> Value;
+        false -> Default
+    end.
+
+send_telemetry(Data, Signature, Id, Height, Version, URL) ->
+    {ok, {Status, _Body}} = httpc:request(
+            post,
+            {URL,
+                    [
+                        {"X-Telemetry-Signature", ?BIN_TO_B58(Signature)},
+                        {"X-Telemetry-Miner-Id", Id},
```

Comment:
I see you're also collecting the name but not sending it in a header?

Reply:
Yeah I wanted a key to use that would be subject to collisions when storing the payloads